File: app/src/test/java/com/productivity/habits/service/TimerStateHolderTest.kt
package com.productivity.habits.service

import com.google.common.truth.Truth.assertThat
import org.junit.Before
import org.junit.Test

class TimerStateHolderTest {

    @Before
    fun setup() {
        TimerStateHolder.stop()
    }

    @Test
    fun `start initializes timer state with running status`() {
        TimerStateHolder.start("habit_1", "Focus Work", 25.0)

        val state = TimerStateHolder.timerState.value
        assertThat(state.habitId).isEqualTo("habit_1")
        assertThat(state.habitTitle).isEqualTo("Focus Work")
        assertThat(state.totalSeconds).isEqualTo(25 * 60L)
        assertThat(state.remainingSeconds).isEqualTo(25 * 60L)
        assertThat(state.status).isEqualTo(TimerStatus.RUNNING)
        assertThat(state.isRunning).isTrue()
    }

    @Test
    fun `pause and resume toggle status appropriately`() {
        TimerStateHolder.start("habit_1", "Focus Work", 25.0)
        TimerStateHolder.pause()

        assertThat(TimerStateHolder.timerState.value.status).isEqualTo(TimerStatus.PAUSED)
        assertThat(TimerStateHolder.timerState.value.isPaused).isTrue()

        TimerStateHolder.resume()
        assertThat(TimerStateHolder.timerState.value.status).isEqualTo(TimerStatus.RUNNING)
    }

    @Test
    fun `adjustRemaining adds and subtracts delta seconds`() {
        TimerStateHolder.start("habit_1", "Focus Work", 25.0)
        TimerStateHolder.adjustRemaining(300L) // +5m

        assertThat(TimerStateHolder.timerState.value.remainingSeconds).isEqualTo(30 * 60L)

        TimerStateHolder.adjustRemaining(-600L) // -10m
        assertThat(TimerStateHolder.timerState.value.remainingSeconds).isEqualTo(20 * 60L)
    }

    @Test
    fun `tick updates remaining seconds and completes when zero`() {
        TimerStateHolder.start("habit_1", "Focus Work", 10.0)
        TimerStateHolder.tick(50L)

        assertThat(TimerStateHolder.timerState.value.remainingSeconds).isEqualTo(50L)
        assertThat(TimerStateHolder.timerState.value.status).isEqualTo(TimerStatus.RUNNING)

        TimerStateHolder.tick(0L)
        assertThat(TimerStateHolder.timerState.value.remainingSeconds).isEqualTo(0L)
        assertThat(TimerStateHolder.timerState.value.status).isEqualTo(TimerStatus.COMPLETED)
        assertThat(TimerStateHolder.timerState.value.isCompleted).isTrue()
    }

    @Test
    fun `reset restores remaining seconds to total seconds and sets status to IDLE`() {
        TimerStateHolder.start("habit_1", "Focus Work", 25.0)
        TimerStateHolder.tick(60L)
        TimerStateHolder.reset()

        val state = TimerStateHolder.timerState.value
        assertThat(state.remainingSeconds).isEqualTo(25 * 60L)
        assertThat(state.status).isEqualTo(TimerStatus.IDLE)
    }
}

File: test/ui/timer_state_holder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:habit_tracker/ui/detail/controllers/timer_state_holder.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    TimerStateHolder.stop();
  });

  test('start initializes timer state with running status', () {
    TimerStateHolder.start('habit_1', 'Focus Work', 25.0);

    final state = TimerStateHolder.timerState;
    expect(state.habitId, 'habit_1');
    expect(state.habitTitle, 'Focus Work');
    expect(state.totalSeconds, 25 * 60);
    expect(state.remainingSeconds, 25 * 60);
    expect(state.status, TimerStatus.running);
    expect(state.isRunning, isTrue);
  });

  test('pause and resume toggle status appropriately', () {
    TimerStateHolder.start('habit_1', 'Focus Work', 25.0);
    TimerStateHolder.pause();

    expect(TimerStateHolder.timerState.status, TimerStatus.paused);
    expect(TimerStateHolder.timerState.isPaused, isTrue);

    TimerStateHolder.resume();
    expect(TimerStateHolder.timerState.status, TimerStatus.running);
  });

  test('adjustRemaining adds and subtracts delta seconds', () {
    TimerStateHolder.start('habit_1', 'Focus Work', 25.0);
    TimerStateHolder.adjustRemaining(300); // +5m

    expect(TimerStateHolder.timerState.remainingSeconds, 30 * 60);

    TimerStateHolder.adjustRemaining(-600); // -10m
    expect(TimerStateHolder.timerState.remainingSeconds, 20 * 60);
  });

  test('tick updates remaining seconds and completes when zero', () {
    TimerStateHolder.start('habit_1', 'Focus Work', 10.0);
    TimerStateHolder.tick(50);

    expect(TimerStateHolder.timerState.remainingSeconds, 50);
    expect(TimerStateHolder.timerState.status, TimerStatus.running);

    TimerStateHolder.tick(0);
    expect(TimerStateHolder.timerState.remainingSeconds, 0);
    expect(TimerStateHolder.timerState.status, TimerStatus.completed);
    expect(TimerStateHolder.timerState.isCompleted, isTrue);
  });

  test('reset restores remaining seconds to total seconds and sets status to IDLE', () {
    TimerStateHolder.start('habit_1', 'Focus Work', 25.0);
    TimerStateHolder.tick(60);
    TimerStateHolder.reset();

    final state = TimerStateHolder.timerState;
    expect(state.remainingSeconds, 25 * 60);
    expect(state.status, TimerStatus.idle);
  });

  test('setDuration sets total and remaining seconds with IDLE status', () {
    TimerStateHolder.setDuration('habit_1', 'Focus Work', 5.0);

    final state = TimerStateHolder.timerState;
    expect(state.habitId, 'habit_1');
    expect(state.totalSeconds, 5 * 60);
    expect(state.remainingSeconds, 5 * 60);
    expect(state.status, TimerStatus.idle);
  });

  test('setRemainingMinutes updates duration and preserves IDLE status when not running', () {
    TimerStateHolder.setDuration('habit_1', 'Focus Work', 10.0);
    TimerStateHolder.setRemainingMinutes(15);

    final state = TimerStateHolder.timerState;
    expect(state.totalSeconds, 15 * 60);
    expect(state.remainingSeconds, 15 * 60);
    expect(state.status, TimerStatus.idle);
  });

  test('focusModeActive is true when running or paused and false otherwise', () {
    // IDLE state
    expect(TimerStateHolder.timerState.focusModeActive, isFalse);

    // RUNNING state
    TimerStateHolder.start('habit_1', 'Focus Work', 25.0);
    expect(TimerStateHolder.timerState.focusModeActive, isTrue);

    // PAUSED state
    TimerStateHolder.pause();
    expect(TimerStateHolder.timerState.focusModeActive, isTrue);

    // COMPLETED state
    TimerStateHolder.tick(0);
    expect(TimerStateHolder.timerState.focusModeActive, isFalse);

    // After stop
    TimerStateHolder.stop();
    expect(TimerStateHolder.timerState.focusModeActive, isFalse);
  });
}
